### PR TITLE
feat: added permission for rest api

### PR DESCRIPTION
**Pull request by olenagerasimova**
part of #1273 

Added permissions, actions, used them in api endpoints.

### CODE REVIEW

File: src/main/java/com/artipie/api/perms/ApiRepositoryPermission.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+     * Required serial.
+     */
+    private static final long serialVersionUID = 8910976571453906971L;
+
+    /**
+     * Ctor.
+     * @param action Action
+     */
+    public ApiRepositoryPermission(final RepositoryAction action) {
+        super(ApiRepositoryPermission.NAME, action.mask, new RepositoryActionList());
```
Comment (dgarus) [inline src/main/java/com/artipie/api/perms/ApiRepositoryPermission.java:32]: `RepositoryActionList` is always the same, it should be a constant.

File: src/main/java/com/artipie/api/perms/ApiRolePermission.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    /**
+     * Ctor.
+     * @param actions Actions set
+     */
+    public ApiRolePermission(final Set<String> actions) {
+        super(
+            ApiRolePermission.NAME,
+            RestApiPermission.maskFromActions(actions, new RoleActionList()),
```
Comment (dgarus) [inline src/main/java/com/artipie/api/perms/ApiRolePermission.java:43]: `RoleActionList ` is always the same, it should be a constant.

File: src/main/java/com/artipie/api/perms/ApiUserPermission.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+     * Required serial.
+     */
+    private static final long serialVersionUID = -1910976021451236971L;
+
+    /**
+     * Ctor.
+     * @param action Action
+     */
+    public ApiUserPermission(final UserAction action) {
+        super(ApiUserPermission.NAME, action.mask, new UserActionList());
```
Comment (dgarus) [inline src/main/java/com/artipie/api/perms/ApiUserPermission.java:33]: `UserActionList ` is always the same, it should be a constant.

File: src/main/java/com/artipie/api/perms/RestApiPermission.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    private final transient boolean any;
+
+    /**
+     * Ctor.
+     * @param name Permission implementation name
+     * @param mask Permission action mask
+     * @param type Actions type
+     */
+    protected RestApiPermission(final String name, final int mask, final ApiActions type) {
+        super(name);
```
Comment (dgarus) [inline src/main/java/com/artipie/api/perms/RestApiPermission.java:61]: Due to the `name` argument for the same class type is always the same, we could pass the canonical class name to the super class to make the ctor parameter list smaller.

File: src/main/java/com/artipie/api/perms/RestApiPermission.java [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        this.type = type;
+        this.any = this.mask == type.all().mask();
+    }
+
+    @Override
+    public final boolean implies(final Permission permission) {
+        final boolean res;
+        if (permission != null && permission.getClass() == this.getClass()) {
+            final RestApiPermission that = (RestApiPermission) permission;
+            res = (this.mask & that.mask) == that.mask && this.getName().equals(that.getName());
```
Comment (dgarus) [inline src/main/java/com/artipie/api/perms/RestApiPermission.java:72]: Due to the name argument for the same class type always the same, there is no sense checking the `name` equality.

File: src/main/java/com/artipie/api/perms/ApiAliasPermission.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+     * Required serial.
+     */
+    private static final long serialVersionUID = -2910962571451212361L;
+
+    /**
+     * Ctor.
+     * @param action Action
+     */
+    public ApiAliasPermission(final AliasAction action) {
+        super(ApiAliasPermission.NAME, action.mask, new AliasActionList());
```
Comment (dgarus) [inline src/main/java/com/artipie/api/perms/ApiAliasPermission.java:33]: `AliasActionList ` is always the same, it should be a constant.

**Review by dgarus — CHANGES_REQUESTED** [review]
Thanks!
Please look at my comments.

File: src/main/java/com/artipie/api/perms/RestApiPermission.java (hunk 4 above)
Comment (olenagerasimova) [reply src/main/java/com/artipie/api/perms/RestApiPermission.java:61]: hm, it's not allowed to call `this.getClass().getCanonicalName()` before or at the same time with calling super constructor.

**Comment by olenagerasimova** [comment]
@dgarus thanks, corrected

**Review by ChGen — APPROVED** [review]
LGTM

**Review by dgarus — APPROVED** [review]
Thanks!

**Pull request merged**